bus-mapping/src/evm/opcodes/returndatacopy.rs

```diff
@@ -80,53 +67,38 @@ fn gen_returndatacopy_step(
         ),
         (
             CallContextField::LastCalleeReturnDataLength,
-            return_data.len().into(),
+            return_data_len.into(),
         ),
     ] {
         state.call_context_read(&mut exec_step, call_id, field, value);
     }
     Ok(exec_step)
 }
 
-fn gen_copy_steps(
-    state: &mut CircuitInputStateRef,
-    exec_step: &mut ExecStep,
-    src_addr: u64,
-    dst_addr: u64,
-    src_addr_end: u64,
-    bytes_left: u64,
-) -> Result<Vec<(u8, bool)>, Error> {
-    let mut copy_steps = Vec::with_capacity(bytes_left as usize);
-    let src_addr_base = state.call()?.last_callee_return_data_offset;
-    for idx in 0..bytes_left {
-        let addr = src_addr + idx;
-        let value = if addr < src_addr_end {
-            state.call_ctx()?.return_data[(addr - src_addr_base) as usize]
-        } else {
-            unreachable!("return data copy out of bound")
-        };
-        // Read
-        state.push_op(
-            exec_step,
-            RW::READ,
-            MemoryOp::new(state.call()?.last_callee_id, addr.into(), value),
-        );
-
-        // Write
-        copy_steps.push((value, false));
-        state.memory_write(exec_step, (dst_addr + idx).into(), value)?;
-    }
-    Ok(copy_steps)
-}
-
 fn gen_copy_event(
     state: &mut CircuitInputStateRef,
     geth_step: &GethExecStep,
+    exec_step: &mut ExecStep,
 ) -> Result<CopyEvent, Error> {
+    let rw_counter_start = state.block_ctx.rwc;
+
     // Get low Uint64 of offset.
-    let dst_addr = geth_step.stack.nth_last(0)?.low_u64();
-    let data_offset = geth_step.stack.nth_last(1)?.as_u64();
-    let length = geth_step.stack.nth_last(2)?.as_u64();
+    let dst_addr = geth_step.stack.nth_last(0)?;
+    let data_offset = geth_step.stack.nth_last(1)?;
+    let length = geth_step.stack.nth_last(2)?;
+
+    let call_ctx = state.call_ctx_mut()?;
+    let memory = &mut call_ctx.memory;
+    memory.extend_for_range(dst_addr, length);
```

Comment:
the `copy_from` below will automatically resize memory if needed. why call extend_for_range manually here?

Reply:
you mean `memory_updated.copy_from(...` in L96 ?  memory_updated is clone of memory. memory_updated's resize don't change here `memory`